fauna/query/iterator.py

```diff
@@ -0,0 +1,67 @@
+from typing import Iterator, Optional
+
+from fauna.client.client import Client, QueryOptions
+from fauna.query import Query, fql, Page
+
+
+class QueryIterator:
+    """A class to provider an iterator on top of Fauna queries."""
+
+    def __init__(self,
+                 client: Client,
+                 fql: Query,
+                 opts: Optional[QueryOptions] = None):
+        """Initializes the QueryIterator
+
+        :param fql: A string, but will eventually be a query expression.
+        :param opts: (Optional) Query Options
+
+        :raises TypeError: Invalid param types
+        """
+        if not isinstance(client, Client):
+            err_msg = f"'client' must be a Client but was a {type(client)}. You can build a " \
+                        f"Client by calling fauna.client.Client()"
+            raise TypeError(err_msg)
+
+        if not isinstance(fql, Query):
+            err_msg = f"'fql' must be a Query but was a {type(fql)}. You can build a " \
+                       f"Query by calling fauna.fql()"
+            raise TypeError(err_msg)
+
+        self.client = client
+        self.fql = fql
+        self.opts = opts
+        self.cursor = None
+
+    def __iter__(self) -> Iterator:
+        return self.iter()
+
+    def iter(self) -> Iterator:
```

Comment:
Let's imply this is private

```suggestion
    def _iter(self) -> Iterator:
```